### ISSUE TITLE
slow rendering of a map

### ISSUE THREAD

**Issue by rinigus** [Karry/osmscout-sailfish#8]
I am not sure its a right time (in terms of development) to report this issue. However, its probably a right channel (github) to get your point of view on that.

First, thank you very much for developing this program - it surely has a great use!

So, on Nexus 4, SFOS 2.0.2.48, I get a map rendered at about 12 seconds on zoom level 16 (default) in rather sparsely populated area. Is this an expected result?

I used a git repo clone from today to check if its fixed already. Looked a bit into the source, tried few things in the options for db, but not much success. 

Best wishes,

rinigus 

PS: please let me know if you need any details or this is a common issue.

**Comment by Karry** [comment]
Well. This topic was discussed many times on [upstream library mailing list](https://sourceforge.net/p/libosmscout/mailman/libosmscout-development/thread/139FFB5E-A890-43C5-BD8E-3AE74DF1C005%40zedsaid.com/#msg35266079). 12 seconds is really too long. It tooks 4 seconds  for me usually (initial rendering with empty caches is little bit slower).

For finding source of problem, you should start OSMScout from terminal (ssh) to see stdout log and grep DrawTileMap method output. 
```
$/usr/bin/harbour-osmscout 2>&1 | grep DrawTileMap
[D] DBThread::DrawTileMap:458 - prepare:    0 
[D] DBThread::DrawTileMap:463 - lookup:     1 
[D] DBThread::DrawTileMap:474 - load data:  4506 
[D] DBThread::DrawTileMap:480 - convert:    678 
[D] DBThread::DrawTileMap:497 - draw:       2713 
```
Printed times are in milliseconds. Or you can inspect journal log (as root) alternatively.
```
journalctl | grep DrawTileMap
```

Rendering consist from three main steps: **data loading** from database (it is IO bounded, caches can helps), **converting** raw data to renderable objects, and main **drawing**.

When data loading is slow, there are just a few thinks to do. Increase caches, inspect data seeks... Or use faster memory (SD card?).
When rendering is slow... There is place for code improvements in library. Slowest part of rendering is  label rendering. This was discussed many times too, but noone has time to reimplement it :-(

**Comment by rinigus** [comment]
Thanks for getting back to me. Yes, I looked into it today and it seems to be general problem. However, maybe there is something I can do to optimize. 

I did make a log and its at https://ptpb.pw/Tlg2.txt/ . Here, I start an app, render a location in Estonia at 16x, and close it. Looks to me that the path rendering is an issue:

Draw: [59.42832 N 24.73572 E - 59.44228 N 24.75220 E] 65536x/16 1152x1920 144.00000 DPI
[D] osmscout::MapPainterQt::HasIcon:110 - Loaded image ' "/usr/share/harbour-osmscout/map-icons//bus_stop.png" ' 
[D] osmscout::MapPainterQt::HasIcon:110 - Loaded image ' "/usr/share/harbour-osmscout/map-icons//parking.png" ' 
Paths: 4582/1426/1426/213 (pcs) 0.056/0.299/0.000/10.428 (sec)
Areas: 3774/1585/3170 (pcs) 0.149/0.144/0.236 (sec)
Nodes: 3185+0/3185 (pcs) 0.339/0.000 (sec)
Labels: 979/4/983 (pcs) 0.357 (sec)

Few replies to your questions: 
* map and everything else  is on internal storage [nexus 4 does not have SD card option];
* fact that you get it in 4 seconds sounds promising :)

I looked a bit in your code and saw that you cache QPixmap's in the TileCache. I'm developing an app that internally also generates images and shows them. I compared 3 options: store cache as QImage, PNG files in /tmp, or QPixmap. For the same cache in terms of images keeping them as PNGs and loading in QML was by far the most memory efficient solution. As for CPU usage, I could not see much difference. Although, in my case, PNGs were generated anyway by external program, so CPU info in my app is not representative. 

I would expect that if the cache is stored on storage as PNGs, there will be significant reduction in required generation of the tiles. While it will not resolve the issue at start, it should help while moving the map around and getting back. I think, that instead of 50, it should be possible to keep 100s or 1000s images in cache. With the present implementation of keeping cache as QPixmap, I could see that RSS memory usage could easily hit 200MB at browsing maps. So, increasing QPixmap based cache by 10x is probably not really an option.

**Comment by rinigus** [comment]
One question: is the rendering multithreaded? It seems to me that only one CPU core is really used with 2-3 of them standing still or switched off. In addition, working CPU core sits only up to 50% of time at max Hz, the rest of the time at much lower frequencies. 

To debug it further, how many working threads I should expect while moving around on a map at 16x zoom?

**Comment by rinigus** [comment]
I am sure now that there is only one thread rendering the map while I am browsing it. As a result, CPU frequency goes to lower values a lot. One working thread is visible using "top -H" as well with couple  QSGRenderThread active, but low CPU%. Is there a way to enable tile generation by multiple threads? That would also bring CPU usage and load higher allowing mpdecision to switch on cores and require higher CPU frequency.

Below are the graphs showing CPU frequency distribution and CPU usage by osmscout. I was browsing a map before and after 14:50.

On the graphs with frequencies, all used frequencies are recorded for all CPU cores (if a core its switched off, then % is zero). I divide usage by number of cores. So, for 4 core CPU, to get 100% I'd need to have all cores working on particular frequency. As you could see, max frequency is used only upto 15% and mostly only one CPU core was working (total sum of all used frequencies is ~25%). 

On the figure showing osmscout stats, CPU usage is ~95%. In this collectd plugin data is shown a bit differently. On that graph, it should be 100% for single thread and, as far as I remember, goes up over n*100% where n is the number of CPUs if they are all used. So, again, while we have many threads (quite large number is usually QML built), only one is really active.

I wonder, is there a setting where I could specify number of active worker threads? I think that by crunching this number to 4 or higher, I should get much faster rendering (if libosmscout supports it).


![screenshot-16-09-04-15-21-28](https://cloud.githubusercontent.com/assets/15706605/18230981/00ce2ae8-72b4-11e6-9860-314972fad93c.png)
![screenshot-16-09-04-15-20-55](https://cloud.githubusercontent.com/assets/15706605/18230982/07040982-72b4-11e6-9af4-f81175c2a2bb.png)

**Comment by rinigus** [comment]
Reading from mailing list: it seems that there is no threading support [yet]. Or am I wrong?

**Comment by Karry** [comment]
**To your multithreading questions:**

OSMScout use one dedicated thread for manipulating with database and rendering of map tiles (`DBThread` object). It helps to keep UI responsive, because UI thread is not blocked (for long time). Core library support multithreading for data loading (one asynchronous request for each database file), but it has minor gain when loading is IO bounded (and it is for usual devices).

Rendering have to be done in single thread. But you can use more workers when every worker has own private painter (`MapPainterQt`). OSMScout for Sailfish merge tile requests with same zoom level before rendering ([DBThread.cpp:834](https://github.com/Karry/osmscout-sailfish/blob/master/src/DBThread.cpp#L834)). It is faster than render each tile separately. 

When you open map, it start rendering initial request, when you zoom out before first request is done, it emits second request that have to wait for first one. This scenario can be improved to use more rendering workers...

**Back to your logs:**

It is clear that database loading is not problem in your case. The worst part is rendering of way labels.

> Paths: 4582/1426/1426/213 (pcs) 0.056/0.299/0.000/**10.428** (sec)

(See (MapPainter.cpp:2136)[https://github.com/Framstag/libosmscout/blob/master/libosmscout-map/src/osmscout/MapPainter.cpp#L2136]) It tooks 10 seconds! It gives me another questions: how old is your export? Do you have contour lines in your map?

Map painter has performance problems with too long ways. For this reason, Tim (author of library) added limit for maximum length of merged ways during import process some weeks ago. Can you try to regenerate your database with updated Import tool?

**Comment by rinigus** [comment]
Maybe the multithreading we can discuss in a separate issue or later, if you don't mind. I am sure that using multithreading would allow to use the device hardware better and would allow all CPU cores to render images faster and not drop to low CPU frequencies, as it is in my device. I guess, its an optimization of merging tiles and rendering in parallel.

The map that I am using is new, I generated it last week (last weekmap and last week source). To be sure, I pulled the last version today from upstream (https://github.com/Framstag/libosmscout) and downloaded the new version from geofabrik.de. 

Yesterday I tested DrawMapCairo and Qt version with them both producing similar result. Today, on the other system where I don't have access to DISPLAY, I used DrawMapCairo for testing. I made small changes in demo code to be as close as possible to the code on phone: 
* DPI was set to 144
* magnification was set to using magnification.SetLevel(zoom) with zoom set to 16
* drawParameter.SetDebugPerformance(true);
* I used osmscout-sailfish/stylesheets/standard.oss

The result was the same for "old" and "new" map:

```
Draw: [59.42832 N 24.73572 E - 59.44228 N 24.75220 E] 65536x/16 1152x1920 144.00000 DPI
ERROR while loading image 'bus_stop'
ERROR while loading image 'parking'
ERROR while loading image 'parking'
Paths: 4639/1429/1429/215 (pcs) 0.003/0.057/0.000/0.621 (sec)
Areas: 2672/1585/3170 (pcs) 0.006/0.049/0.063 (sec)
Nodes: 3190+0/3190 (pcs) 0.102/0.000 (sec)
Labels: 796/4/800 (pcs) 0.042 (sec)

real    0m1.393s
user    0m1.394s
sys     0m0.089s
```

Whether the map has contours or not, I don't know. How can I check it? The import log is at https://ptpb.pw/Q96Q.txt .

![test_map_new](https://cloud.githubusercontent.com/assets/15706605/18240688/f35f8ba2-7355-11e6-8a29-fef5b3c52837.png)

The rendered image is pasted below. What is the blue line in the middle (top to center), I don't know. It appears sometimes.

**Comment by rinigus** [comment]
PS: If you wish, I can upload the map somewhere. Or maybe you have a reference region and I could convert maps and test on that?

**Comment by Karry** [comment]
It seems that there is something wrong with OSM data, probably some long way that is missing in some imported index cell... I will try this to inspect later today...

**Comment by rinigus** [comment]
Thank you very much! I presume its then covering the whole map, since the rendering problem is all over Estonia, as far as I could tell. Thank you again for fast reply and support!

**Comment by rinigus** [comment]
I have just checked on a map of Brussels - the rendering is much faster and that particular line in the logs would be 

Paths: 8526/1990/1990/627 (pcs) 0.102/0.353/0.004/0.440 (sec)

for the same magnification. So, now its very interesting to know what's particularly bad in that map file of Estonia...

**Comment by Karry** [comment]
My import is done. Good message is that I am able to reproduce it :) Now, we should print times for way labeling. Then find what way(s) cause problems. Then, when you know what way is problematic, you can print its file offset. It can be used to obtain OSM id (`DumpData` tool)... Then we should see if this way is correct or should be fixed in OSM database...

**Comment by rinigus** [comment]
Excellent news!

Since I am very new to all this OSM business [and maps overall], please drive me through it. I guess the easiest is to do it on desktop and use the demo programs for it, if possible.

How do I print way labeling times?

**Comment by Karry** [comment]
I added some prints to `MapPainter::DrawWayContourLabels` in this commit: https://github.com/karry-space-with-my-second-forks/libosmscout/commit/f05c658db531fda4878e556e5d346545ab0facd7

Now, it is clear from log:
```
way offset: 10812284 labeling time: 12.070
```
This simple way with 158 nodes tooks 12 seconds to render. Its OSM id...
```
$ ./DumpData/DumpData /media/data/tmp/scout-map/estonia -wo 10812284    
Way {
  OSM id: 128743565
  fileOffset: 10812284
  type: route_ferry
  boundingBox: [54.53199 N 18.50292 E - 60.22047 N 25.22533 E]
  center: 57.37623 N 21.86412 E
  cell level: 4

  Name: Rostock-Gdynia-Helsinki
...
```

Damn! It is valid way - it is ferry line to Helsinky: http://www.openstreetmap.org/way/128743565
It is known problem that have to be solved in upstream library... Now, as simple workaround, you can disable labels for ferry in standard.ost stylesheet. 

As final solution, we have to implement some systematic method. 
 - Split long ways in import process?
 - Build special index for such ways?
 - Optimize labeling code?

What do you think Tim? (ping @Framstag )

**Comment by rinigus** [comment]
just to confirm, we are down to sub-second values:

Paths: 4582/1426/1426/212 (pcs) 0.098/0.386/0.000/0.195 (sec)

I just removed route_ferry from https://github.com/Karry/osmscout-sailfish/blob/master/stylesheets/standard.oss#L1023

So, rendering has improved a lot by this fix, thank you very much for looking into it.

Now, should we be just happy or do you want to look into multithreading aspects? I would expect faster rendering and better use of a hardware if its enabled. While moving around, I am still mainly using average frequencies on a single CPU core and wait for the map to be rendered, which is suboptimal. Or maybe I should look into it a bit more and discuss it separately later?

Thanks for this workaround!

**Comment by Karry** [comment]
You're welcome. 

Well, when you have only one request in queue, there is no space for multithread processing. Multiple requests can be processed in parallel of course, but I want to focus to another topics now (map download manager + download server, optimise rendering of huge areas and long ways...)

I mean that merge requests are always welcome and I am prepared to help you with that. But I have another priorities now.

**Comment by rinigus** [comment]
No problem! Sure, thank you for your program and support!

rinigus

**Comment by Framstag** [comment]
Hello, I'm the main author of libosmscout. Thanks for this very interesting discussion and in deep analysis. 

As author of libosmscout I`m very interested in on device performance analysis. As such label drawing was identified as current main problem during rendering. I did improve parts of this though there are still problems. https://sourceforge.net/p/libosmscout/mailman/libosmscout-development/thread/20160904122933.Horde.MMZWIBTDHsaKyKFmNCI3Rw2%40webmail.df.eu/#msg35335865 summarizes the current state.

From the analysis is sounds like the actual problem is contour labels for route_ferry (the standard style sheet would draw them starting with zoom level 13). Problem with long ways is the currently in some edge cases - especially for long ways - inefficient clipping code. For detection if a contour labels has to be drawn the bounding box of the visible area and of the way are compared for intersection. If there is intersection all contour labels along the way are draw.

In worst case ways for example have some "U" shape where the visible area is completely within the middle area. So in fact there is no intersection though the bounding boxes do intersect (other shapes are possible of course).

Drawing contour labels is backend specific so each backend (in this case Qt) has to find its own solution though the principle solution is likely the same.

In the Qt version each letter is rotated and draw separately (see https://github.com/Framstag/libosmscout/blob/master/libosmscout-map-qt/src/osmscout/MapPainterQt.cpp, line 399 ff.). So you could do clipping during drawing of each letter. I'm not sure if this would gain much speed, because there are many letters ;-) Alternative solution would be to calculate the intersections with the visible area and during drawing identify the parts that are in and the parts that are out. The problem in both solutions is that the way is not just a line but a polygon consisting of many lines, which makes handling a little bit trickier. Btw. drawing contour symbol has the same problems.

I'm open to any simple though optimal geometric solution :-) I'm not sure that making way shorter or using a different presentation should be used. Such tricks will have negative impact on on disk size, mapping between OSM and libosmscout, routing, original data is lost ...So I wold like to avoid them. We do have same special casing for coastlines but this comes at its costs.

Regarding multithreading: As stated multithreading is used for fetching data from various indexes at the same time. Since the threads mainly trigger IO and IO may even saturate if done from multiple threads you will not get full CPU load for this.

For actual map rendering no mutithreading during rendering of *one* image is currently used. I'm currently under the assumption that most of the GUI engine do not support this (so some may), because they have (hidden) state. Also rendering has some certain order in it self so you cannot parallelize everything anyway. There are though some parts within the rendering code that could get parallelized like sorting objects. It is also possible that some parts like getting text dimensions are parallelizable if using separate OS handles. I'm not sure how much this will improve drawing.

So the main gain would be to rendering *different* images in parallel. For this you would instantiate multiple MapPainters and let the render into different bitmaps. This should work, if not, it is something that should be fixed. Multiple MapPainter will though increase memory footprint.

**Comment by rinigus** [comment]
I can understand that having many exceptions are problematic. I presume that particular ferry line should influence a lot of users along the cost in EU. If I do understand correctly, the issue came out because of the label of this line. If true, naïve solution would be: maybe so long lines should not warrant a label if they are out of the region? For example not within few region spans? But I am sure there are reasons why its a terrible idea :)

Maybe, for now, the ferry line names should not be shown on mobile applications? I am sure many in sweden & germany may benefit from it.

As for multithreading: I'll have to think and use osmscout more. After fixing this issue, the speed of rendering became so much faster that I am not bothered with little wait. I'll probably try to make some small project to learn around it and hopefully I can send some ideas/PRs back. But it would take some time...

**Comment by Framstag** [comment]
In principle the problem can occur for any way or area type where ar specific instance is "big". So it can help in a concrete case to patch the style sheet but there will be other cases.

Why are way long? For data size and also for routing speed libosmscout is interested in having long ways. Long ways mean less data overhead and a simple routing graph.

For label rendering though is important to only render labels that are actually visible. Short ways solve this problem better than longer ways (though short ways do not solve the problem algorithmically).

IF you are interested in libosmscout: you are of course welcome there are a number of examples and and increasing amount fo documentation that should help you.

**Comment by Karry** [comment]
Thank @rinigus and @Framstag for investigate this issue. It seems that rendering is "fast enough" with current codebase and Qt frontend. I am closing this old issue.
